akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -473,6 +473,177 @@ private[akka] final case class Conflate[In, Out](seed: In ⇒ Out, aggregate: (O
+        val elem = grab(in)
+        val cost = costFn(elem)
+        if (agg == null) {
+          left -= cost
+          agg = seed(elem)
+        } else if (left < cost) {
+          pending = elem
+        } else {
+          left -= cost
+          agg = aggregate(agg.asInstanceOf[Out], elem)
+        }
+
+        if (isAvailable(out)) flush()
+        if (pending == null) pull(in)
+      }
+
+      override def onUpstreamFinish(): Unit = {
+        if (agg == null) completeStage()
+      }
+    })
+
+    setHandler(out, new OutHandler {
+
+      override def onPull(): Unit = {
+        //if upstream finished, we still might emit up to 2 more elements (whatever agg is + possibly a pending heavy element)
+        if (isClosed(in)) {
+          if (agg == null) completeStage()
+          else {
+            push(out, agg.asInstanceOf[Out])
+            if (pending == null) completeStage()
+            else {
+              agg = seed(pending.asInstanceOf[In])
+              pending = null
+            }
+          }
+        } else if (agg == null) {
+          if (!hasBeenPulled(in))
+            pull(in)
+        } else {
+          flush()
+          if (!hasBeenPulled(in)) pull(in)
+        }
+      }
+    })
+  }
+}
+
+private[akka] final case class BatchWeighted[I, O](max: Long, costFn: I ⇒ Long, seed: I ⇒ O, aggregate: (O, I) ⇒ O)
+  extends AbstractBatch(max, costFn, seed, aggregate, Inlet[I]("BatchWeighted.in"), Outlet[O]("BatchWeighted.out")) {
+  override def initialAttributes = Attributes.name("BatchWeighted")
+}
+
+private[akka] final case class Batch[I, O](max: Long, seed: I ⇒ O, aggregate: (O, I) ⇒ O)
+  extends AbstractBatch(max, { _: I ⇒ 1L }, seed, aggregate, Inlet[I]("Batch.in"), Outlet[O]("Batch.out")) {
+  override def initialAttributes = Attributes.name("Batch")
+}
+
+//private[akka] final case class Conflate[I, O](seed: I ⇒ O, aggregate: (O, I) ⇒ O)
+//  extends AbstractBatch(Long.MaxValue, { _: I ⇒ 0L }, seed, aggregate, Inlet[I]("Conflate.in"), Outlet[O]("Conflate.out")) {
+//  override def initialAttributes = Attributes.name("Conflate")
+//}
+
+/**
+ * INTERNAL API
+ */
+private[akka] final case class AggregateWeighted[In, Out](max: Long, costFn: In ⇒ Long, seed: In ⇒ Out,
+                                                          aggregate: (Out, In) ⇒ Out,
+                                                          decider: Supervision.Decider) extends DetachedStage[In, Out] {
+  private var agg: Any = null
+  private var left: Long = max
+  private var pending: Any = null
+
+  private[this] def flush(ctx: DetachedContext[Out]) = {
+    val result = agg.asInstanceOf[Out]
+    agg = null
+    left = max
+    if (pending != null) {
+      val elem = pending.asInstanceOf[In]
+      agg = seed(elem)
+      left -= costFn(elem)
+      pending = null
+    }
+    ctx.pushAndPull(result)
+  }
+
+  override def onPush(elem: In, ctx: DetachedContext[Out]): UpstreamDirective = {
+    val cost = costFn(elem)
```

Comment:
For aggregating ByteStrings this is appropriate, but when compressing multiple sets by performing a union it would be more interesting to apply the cost function to the resulting aggregated value. @drewhk WDYT?